site/src/api/queries/groups.ts

```diff
@@ -25,6 +25,48 @@ export const group = (groupId: string) => {
   };
 };
 
+export type GroupsByUserId = Readonly<Map<string, readonly Group[]>>;
+
+export function groupsByUserId(organizationId: string) {
+  return {
+    ...groups(organizationId),
+    select: (allGroups) => {
+      const userIdMapper = new Map<string, Group[]>();
+
+      for (const group of allGroups) {
+        for (const user of group.members) {
+          let groupsForUser = userIdMapper.get(user.id);
+          if (groupsForUser === undefined) {
+            groupsForUser = [];
+            userIdMapper.set(user.id, groupsForUser);
+          }
+
+          groupsForUser.push(group);
+        }
+      }
+
+      // Defined outside the loop because it can be reused by all group arrays,
+      // and doesn't need to be rebuilt from scratch on each iteration
+      const orderGroupsByName = (g1: Group, g2: Group) => {
+        const key =
+          g1.display_name && g2.display_name ? "display_name" : "name";
+
+        if (g1[key] === g2[key]) {
+          return 0;
+        }
+
+        return g1[key] < g2[key] ? -1 : 1;
+      };
+
+      for (const groupsList of userIdMapper.values()) {
+        groupsList.sort(orderGroupsByName);
+      }
+
+      return userIdMapper as GroupsByUserId;
+    },
+  } satisfies UseQueryOptions<Group[], unknown, GroupsByUserId>;
+}
```

Comment:
We use lodash so maybe it can be helpful to simplify some of these logic.

Reply:
Yeah, I've never used Lodash before, so to be honest, I keep forgetting that we have it, but I'll see if some of their utilities can help out

Reply:
Tried going through the documentation, but it didn't seem like Lodash has support for Map objects, and I didn't want to "downgrade" the type by turning it into a regular JS object (which would mean losing better memory usage, easy iteration, and easy size checking)

I did refactor the code to consolidate some loops, though